src/lib/expression/evaluation/expression_evaluator.cpp

```diff
@@ -0,0 +1,1054 @@
+     * Note that other DBMS behave differently when it comes to negative indices.
+     *
+     * START -8 -7 -6 -5 -4 -3 -2 -1 || 0  || 1 2 3 4 5 6  7  8
+     * CHAR  // // // H  e  l  l  o  || // || H e l l o // // //
+     *
+     * SUBSTR('HELLO', 0, 2) -> 'H'
+     * SUBSTR('HELLO', -1, 2) -> 'O'
+     * SUBSTR('HELLO', -8, 1) -> ''
+     * SUBSTR('HELLO', -8, 5) -> 'HE'
+     */
+    auto end = int32_t{0};
+    if (start < 0) {
+      start += signed_string_size;
+    } else {
+      if (start == 0) {
+        length -= 1;
+      } else {
+        start -= 1;
+      }
+    }
+
+    end = start + length;
+    start = std::max(0, start);
+    end = std::min(end, signed_string_size);
+    length = end - start;
+
+    // Invalid/out of range arguments, like SUBSTR("HELLO", 4000, -2), lead to an empty string
+    if (!string.empty() && start >= 0 && start < signed_string_size && length > 0) {
+      length = std::min<int32_t>(signed_string_size - start, length);
+      result_values[chunk_offset] = string.substr(static_cast<size_t>(start), static_cast<size_t>(length));
+    }
+  }
+
+  return std::make_shared<ExpressionResult<std::string>>(result_values, result_nulls);
+}
+
+std::shared_ptr<ExpressionResult<std::string>> ExpressionEvaluator::_evaluate_concatenate(
+    const std::vector<std::shared_ptr<AbstractExpression>>& arguments) {
+  std::vector<std::shared_ptr<ExpressionResult<std::string>>> argument_results;
+  argument_results.reserve(arguments.size());
+
+  auto result_is_nullable = false;
+
+  // I - Compute the arguments
+  for (const auto& argument : arguments) {
+    // CONCAT with a NULL literal argument -> result is NULL
+    if (argument->data_type() == DataType::Null) {
+      auto null_value_result = ExpressionResult<std::string>{{{}}, {true}};
+      return std::make_shared<ExpressionResult<std::string>>(null_value_result);
+    }
+
+    const auto argument_result = evaluate_expression_to_result<std::string>(*argument);
+    argument_results.emplace_back(argument_result);
+
+    result_is_nullable |= argument_result->is_nullable();
+  }
+
+  // II - Compute the number of output rows
+  auto result_size = argument_results.empty() ? size_t{0} : argument_results.front()->size();
+  for (auto argument_idx = size_t{1}; argument_idx < argument_results.size(); ++argument_idx) {
+    result_size = _result_size(result_size, argument_results[argument_idx]->size());
+  }
+
+  // II - Concatenate the values
+  std::vector<std::string> result_values(result_size);
+  for (const auto& argument_result : argument_results) {
+    argument_result->as_view([&](const auto& argument_view) {
+      for (auto chunk_offset = ChunkOffset{0}; chunk_offset < result_size; ++chunk_offset) {
+        // The actual CONCAT
+        result_values[chunk_offset] += argument_view.value(chunk_offset);
+      }
+    });
+  }
+
+  // III - Optionally concatenate the nulls and return
```

Comment:
What does "concatenating nulls" mean?

Reply:
Ah, is it about `SELECT CONCAT('hallo', 'moritz', NULL)`?
MySQL gives me NULL for this, Postgres `hallomoritz`. We should clarify what we do.

Reply:
Added `(i.e. one argument is null -> result is null)` Let's discuss `CONCAT('hallo', 'moritz', NULL)` in person and establish a policy for null handling in future functions